snowdrop-execution/src/Snowdrop/Execution/DbActions/AVLp/Actions.hs

```diff
@@ -16,77 +17,86 @@ import qualified Data.Tree.AVL as AVL
+                                                          clientModeToTempSt, runAVLCacheT)
 import           Snowdrop.Execution.DbActions.Types (ClientMode (..), DbAccessActions (..),
                                                      DbAccessActionsM (..), DbAccessActionsU (..),
                                                      DbActionsException (..), DbModifyActions (..),
                                                      RememberForProof (..))
-import           Snowdrop.Util (HasPrism, NewestFirst, inj, proj)
+import           Snowdrop.Util (HKey, HVal, HasPrism, NewestFirst, inj, proj, unHSetEl)
 
 avlClientDbActions
-  :: forall k v m h n undo.
-    ( KVConstraint k v
-    , MonadIO m
-    , MonadCatch m
+    :: forall h xs undo m n .
+    ( MonadIO m, MonadCatch m, MonadIO n, MonadCatch n
     , AvlHashable h
-    , AVL.Hash h k v
-    , MonadIO n
-    , MonadCatch n
-    , RetrieveImpl (ReaderT (ClientTempState h k v n) m) h
-    , Serialisable (MapLayer h k v h)
-    , HasPrism undo (AvlUndo h)
+    , RetrieveImpl (ReaderT (ClientTempState h xs n) m) h
+    , AllAvlEntries h xs
+    , HasPrism undo (AvlUndo h xs)
     )
     => RetrieveF h n
-    -> RootHash h
-    -> n (ClientMode (AVL.Proof h k v) -> DbModifyActions (AVLChgAccum h k v) undo k v m ())
+    -> RootHashes h xs
+    -> n (ClientMode (AvlProofs h xs) -> DbModifyActions (AVLChgAccums h xs) undo xs m ())
 avlClientDbActions retrieveF = fmap mkActions . newTVarIO
   where
+    mkActions
+        :: TVar (RootHashes h xs)
+        -> ClientMode (AvlProofs h xs)
+        -> DbModifyActions (AVLChgAccums h xs) undo xs m ()
     mkActions var ctMode =
-        DbModifyActions (mkAccessActions var ctMode) (reThrowAVLEx @k . apply var)
+        DbModifyActions (mkAccessActions var ctMode) (apply var)
 
+    mkAccessActions
+        :: TVar (RootHashes h xs)
+        -> ClientMode (AvlProofs h xs)
+        -> DbAccessActionsU (AVLChgAccums h xs) undo xs  m
     mkAccessActions var ctMode = daaU
       where
         daa =
           DbAccessActions
             -- adding keys to amsRequested
-            (\cA req -> reThrowAVLEx @k $ query cA req =<< createState)
+          (\cA req -> createState >>= \ctx -> query ctx cA req)
             -- setting amsRequested to AMSWholeTree as iteration with
             -- current implementation requires whole tree traversal
-            (\cA p b f -> reThrowAVLEx @k $ iter cA p b f =<< createState)
-        daaM = DbAccessActionsM daa (\cA cs -> reThrowAVLEx @k $ modAccum cA cs =<< createState)
+          (\cA -> createState >>= \ctx -> iter ctx cA)
+        daaM = DbAccessActionsM daa (\cA cs -> createState >>= \ctx -> modAccum ctx cA cs)
         daaU = DbAccessActionsU daaM
                   (withProjUndo . modAccumU)
-                  (\cA _cs -> pure . Right . inj . computeUndo cA =<< reThrowAVLEx @k createState)
+                  (\cA _cs -> pure . Right . inj . computeUndo cA =<< createState)
```

Comment:
Function `createState` may also throw an exception, thus must be wrapped into `reThrowAVLEx`

Reply:
>Function createState may also throw an exception

how? `clientModeToTempSt` can't because there is such line
```
convert (AvlProof p@(AVL.Proof avl)) rootH cache = reThrowAVLEx @(HKey r) @h $ do
```

Reply:
I think I added this line after I added the comment (while debugging block sync)